lib/puppet/module/task.rb

```diff
@@ -61,6 +62,58 @@ def self.is_tasks_filename?(path)
       return true
     end
 
+    def self.get_file_details(path, name_start)
+      unless File.absolute_path(path) == File.path(path)
+        msg = _("File pathnames cannot include relative paths")
+        raise InvalidMetadata.new(msg, 'puppet.tasks/invalid-metadata')
+      end
+
+      # This gets the path from the starting point onward
+      # For files this should be the file subpath from the metadata
+      # For directories it should be the directory subpath plus whatever we globbed
+      name = "#{name_start}#{path.partition(name_start).last}"
+
+      { "name" => name, "path" =>  path }
+    end
+
+    # Find task's required lib files and retrieve paths
+    # for both 'files' and 'implementation:files' metadata keys
+    def self.find_files(files, mod)
+      env = mod.environment.respond_to?(:name) ? mod.environment.name : 'production'
+
+      file_list = files.flat_map do |file|
+        module_name, mount, endpath = file.split("/", 3)
+
+        pup_module = Puppet::Module.find(module_name, env)
+        if pup_module.nil?
+          raise Puppet::Module::MissingModule, _("Module %{module_name} not found in environment %{environment_name}.") %
+            {module_name: pup_module.name, environment_name: env}
+        end
+
+        unless MOUNTS.include? mount
+          msg = _("Files must be saved in module directories that Puppet makes available via mount points: %{mounts}" % 
+                  {mounts: MOUNTS.join(', ')})
+          raise InvalidMetadata.new(msg, 'puppet.tasks/invalid-metadata')
+        end
+
+        path = File.join(pup_module.path, mount, endpath)
+        unless File.exist?(path)
+          msg = _("Could not find %{path} on disk" % { path: path })
+          raise InvalidFile.new(msg)
+        end
+
+        if File.directory?(path)
```

Comment:
If path is a directory, then we should error if `file` doesn't include a trailing slash. Similarly, if path is not a directory, we should error if `file` includes a trailing slash.